### ISSUE TITLE
"Variables may not be used here" for `prevent_destroy`

### ISSUE THREAD

**Issue by tomasaschan** [hashicorp/terraform#22544]
### Terraform Version

```
Terraform v0.12.6
```

### Terraform Configuration Files

```hcl
locals {
  test = true
}

resource "null_resource" "res" {
  lifecycle {
    prevent_destroy = locals.test
  }
}

terraform {
  required_version = "~> 0.12.6"
}
```

### Steps to Reproduce

1. `terraform init`

### Description

The [documentation](https://www.terraform.io/docs/configuration/resources.html#lifecycle-lifecycle-customizations) notes that 

> [...] only literal values can be used because the processing happens too early for arbitrary expression evaluation.

so while I'm bummed that this doesn't work, I understand that I shouldn't expect it to.

However, we discovered this behavior because running `terraform init` failed where it had once worked. And indeed, if you comment out the variable reference in the snippet above, and replace it with `prevent_destroy = false`, it works - and if you then change it back _it keeps working_.

Is that intended behavior? And will it, if I do this workaround, keep working?

### Debug Output

```
λ terraform init
2019/08/21 15:48:54 [INFO] Terraform version: 0.12.6
2019/08/21 15:48:54 [INFO] Go runtime version: go1.12.4
2019/08/21 15:48:54 [INFO] CLI args: []string{"C:\\Users\\Tomas Aschan\\scoop\\apps\\terraform\\current\\terraform.exe", "init"}
2019/08/21 15:48:54 [DEBUG] Attempting to open CLI config file: C:\Users\Tomas Aschan\AppData\Roaming\terraform.rc
2019/08/21 15:48:54 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2019/08/21 15:48:54 [INFO] CLI command args: []string{"init"}
There are some problems with the configuration, described below.

The Terraform configuration must be valid before initialization so that
Terraform can determine which modules and providers need to be installed.

Error: Variables not allowed

  on main.tf line 7, in resource "null_resource" "res":
   7:     prevent_destroy = locals.test

Variables may not be used here.


Error: Unsuitable value type

  on main.tf line 7, in resource "null_resource" "res":
   7:     prevent_destroy = locals.test

Unsuitable value: value must be known
```

**Comment by teamterraform** [comment]
Hi @tomasaschan,

`prevent_destroy` cannot support references like that, so if you are not seeing an error then the bug is that the error isn't being shown; the reference will still not be evaluated.

**Comment by baurmatt** [comment]
Just ran into this but with a "normal" variable. It would be create if we can use variables in the `lifecycle` block because without using variables I'm literally unable to use `prevent_destroy` in combination with a "Destroy-Time Provisioner" in a module.

**Comment by mleonhard** [comment]
I'm hitting this, too.  Please allow variables derived from static values to be used in `lifecycle` blocks.  This would let me effectively use modules to run dev & test environments with the same config as prod, while providing deletion protection for prod resources.  AWS RDS has a `deletion_protection` option that is easy to set.  S3 Buckets have an `mfa_delete` option which is difficult to enable.  I found no way to prevent accidental deletion of an Elastic Beanstalk Application Environment.

```
module "backend" {
  source = "../backend"
  flavor = "dev"
  ...
}
```

```
resource "aws_elastic_beanstalk_environment" "api_service" {
  lifecycle {
    prevent_destroy = (var.flavor == "prod")  // <-- error
  }
  ...
}
```

**Comment by weldrake13** [comment]
Seen multiple threads like this. There is an ongoing issue (https://github.com/hashicorp/terraform/issues/3116) which is currently open but @teamterraform seem to have made that private to contributors only.
The need to set lifecycle properties as variables is required in a lot of production environments. 
We are trying to give our development teams control of their infrastructure whilst maintaining standards using modules. Deployment is 100% automated for us, and if the dev teams need to make a change to a resource, or remove it then that change would have gone through appropriate testing and peer review before being checked into master and deployed. 

Our modules need to be capable of having lifecycle as variables. 
Can we get an answer as to why this is not supported?

**Comment by sdemura** [comment]
My use case is very much like @weldrake13's. It would be nice to understand why this can't work.

**Comment by sudhanwadindorkar** [comment]
I would also appreciate if Terraform allows variables for specifying "prevent_destroy" values. As a workaround, since we use the S3 backend for managing our Terraform workspaces, I block the access to the Terraform workspace S3 bucket for the Terraform IAM user in my shell script after Terraform has finished creating the prod resources. This effectively locks down the infrastructure in the workspace and requires a IAM policy change to re-enable it.

**Comment by throwaway8787** [comment]
I write tests for my modules. I need to be able to re-run tests over and over. There's no way for me to delete buckets in a test account and set protection in a production account. Swing and a miss on this one.

**Comment by Clete2** [comment]
Is there a general issue open with Terraform to improve conditional support? Off the top of my head I can think of the following limitations:

- Variable defaults / declarations cannot use conditionals
- Lifecycle rules cannot use conditionals
- `provider =` argument cannot use conditionals
- Modules cannot have count set

All of these make writing enterprise-level Terraform code difficult and more dangerous.

**Comment by rios0rios0** [comment]
The same of: https://github.com/hashicorp/terraform/issues/3116
Can you close, please?

**Comment by weldrake13** [comment]
> The same of: #3116
> Can you close, please?

Hashicorp locked down 3116. If this gets closed then those following cant view the issue.

**Comment by hawksight** [comment]
It's over 4 years since #3116 was opened, I think we'd all appreciate some indication of where this is? Is it still waiting on the proposal mentioned in [this comment](https://github.com/hashicorp/terraform/issues/3116#issuecomment-296848322), #4149 ?

Thought I'd offer up a work around I've used in some small cases. Example here is a module for gcloud sql instance, where obviously in production I want to protect it, but more ephemeral environments I want to be able to pull the environment down without editing the code temporarily.

It's not pretty but it works, and is hidden away in the module for the most part:

```hcl
### variables.tf
variable "conf" {
  type = map(object({
    database_version  = string
    ...
    prevent_destroy   = string
  }))
  description = "Map of configuration per environment"
  default = {
    dev = {
      database_version  = "POSTGRES_9_6"
      ...
      prevent_destroy   = "false"
    }
  # add more env configs here
  }
}

variable "env" {
  type        = string
  description = "Custom environment used to select conf settings"
  default     = "dev"
}

### main.tf
resource "google_sql_database_instance" "protected" {
  count   = var.conf[var.env]["prevent_destroy"] == "true" ? 1 : 0
  ...
  lifecycle {
    prevent_destroy = "true"
  }
}

resource "google_sql_database_instance" "unprotected" {
  count = var.conf[var.env]["prevent_destroy"] == "false" ? 1 : 0
  ...
  lifecycle {
    prevent_destroy = "false"
  }
}

### outputs.tf
output "connection_string" {
  value = coalescelist(
    google_sql_database_instance.protected.*.connection_name,
    google_sql_database_instance.unprotected.*.connection_name,
  )
  description = "Connection string for accessing database"
}
```

Module originated prior to 0.12, so those conditionals could well be shortened using `bool` now. Also I appreciate this is one resource duplicated, and it would be much worse elsewhere for larger configurations.

**Comment by realsby** [comment]
It is so funny. I am asking this question WHY? WHY?

**Comment by Penumbra69** [comment]
I know it's been 4 years in the asking - but also a long time now in the replying.  Commenting on #3119 was locked almost 2 years ago saying "We'll open it again when we are working on this".

Can someone with the inner knowledge of this "feature" work please step up and give us some definitive answers on simple things like:

- If this will be done?
- Is it even on your feature/sprint/planning/roadmap or just a item backlog only?
- When may be expected if it IS on the roadmap?

Thanks for your work - Hashicorp - this tool is awesome!  Not slanting at you, just frustrated that this feature is languishing and I NEED it ... Now....

**Comment by danieldreier** [comment]
@Penumbra69 and all the folks on here: I hear you, and the use cases you're describing totally make sense to me. I'm recategorizing this as an enhancement request because although it doesn't work the way you want it to, this is a known limitation rather than an accidental bug.

**Comment by alex-3sr** [comment]
Hi team

Maybe a duplicate of https://github.com/hashicorp/terraform/issues/3116 ?

**Comment by bmurtagh** [comment]
@danieldreier given that Hashicorp has acknowledge this issue as a "known limitation", is the company able to provide a standard or recommended workaround to address this?

**Comment by danpilch** [comment]
I think the recommended workaround is find-and-replace value before running terraform :(

**Comment by richtong** [comment]
> I think the recommended workaround is find-and-replace value before running terraform :(

Wow this is a real problem so either we duplicate all resources with prevent_destroy, you we use m4 or something to do a search for this (like you have to do with Dockerfiles. pretty ugly :-)

**Comment by kennethtxytqw** [comment]
This might not be appetizing... but here is a workaround.
```
  dynamic "lifecycle" {
    for_each = var.prevent_destroy ? [0] : []
    content {
      prevent_destroy = true
    }
  }
```

**Comment by kennethtxytqw** [comment]
Ahh i tried using dynamic didnt work either.
There is another issue that could be similar to this one at https://github.com/hashicorp/terraform/issues/24188

**Comment by Ashtonian** [comment]
makes the prevent_destroy pretty painful to use with a dev/prod env for multiple resources. Would like to use it with an assertion something like: 

```tf
locals {
  is_prod = replace(lower(var.environment), "prod", "") != lower(var.environment)
}

module "prod_lock" {
  source        = "rhythmictech/errorcheck/terraform"
  version       = "1.0.0"
  assert        = local.is_prod == var.prevent_destroy
  error_message = "you are potentially attempting to destroy prod. Please stop."
}

resource "null_resource" "really important" {
  lifecycle {
    prevent_destroy = var.prevent_destroy
  }
  depends_on [
	module.prod_lock
  ]
}

```

**Comment by tostangs** [comment]
I'm in dire need of this functionality support. I'm having to find and replace instead of just setting it from a local or a var. I'm not sure why this is such a point of contention to get working but I'm sure there's more to it than what I can see, so any additional information on the progress or lack thereof would be greatly appreciated!

**Comment by rd-michel** [comment]
we're also looking forward to use variables in lifecycle blocks.

example:

set prevent_destroy to true per default ... overwrite it in our dev environment on a case by case basis

**Comment by aradwyr** [comment]
Does using `deletion_protection` circumvent this issue in restrictive lifecycle blocks? https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/db_instance#deletion_protection

**Comment by Nogginboink** [comment]
Just ran across this as well. How disappointing. Maybe my "+1" comment will encourage Hashicorp to re-evaluate this.

**Comment by kinghuang** [comment]
> Does using `deletion_protection` circumvent this issue in restrictive lifecycle blocks? https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/db_instance#deletion_protection

Hardly. That's a feature of one very specific resource. `prevent_destroy` is applicable to any Terraform resource.

**Comment by iamtito** [comment]
Looks like this is still an issue.

**Comment by sri53** [comment]
I am facing this issue still in 2021

**Comment by flobilosaurus** [comment]
As a possible workaround resources can be duplicated and conditionally created based on a variable:

```terraform
resource "aws_s3_bucket" "example" {
  count = var.prevent_destroy ? 1 : 0
  bucket = "example"

  lifecycle {
    prevent_destroy = true
  }
# ...
}

resource "aws_s3_bucket" "example_no_prevent_destroy" {
  count = var.prevent_destroy ? 0 : 1
  bucket = "example"

  lifecycle {
    prevent_destroy = false
  }
# ...
}
```

**Comment by d8aninja** [comment]
+1 for requested feature and to keep this alive.   

I know it's (probably?) more complicated than it sounds but I almost spit out my coffee when I read `ctrl-f`.

**Comment by Starefossen** [comment]
Wow! This came as a surprise that using a variable for `prevent_destroy` is not supported! Makes complex environments super difficult!

**Comment by vkozyrev** [comment]
+1

**Comment by crw** [comment]
Just a reminder to please use the 👍 reaction on the original post to upvote issues - we do sort by most upvoted to understand which issues are the most important. This also reduces "noise" in the notification feed for folks following this issue. Thanks!

**Comment by danivendetta** [comment]
We have the same needs. For us make sense to have differents lifecycle - prevent_destroy = [true|false]  for differents envirnoments.

@teamterraform 

For example:

```
workspace = production or develop
life_cycle =     {
    production = "true"
    develop    = "false"
  }
prevent_destroy = lookup(local.life_cycle , terraform.workspace)

```
If we're thinking in a very large companys, it's not maintainable to do massive updates in that statment for delete develop or test environments, and undo de change again for set production environment protected. Also, while test environment is deleted, production environment it's unprotected.

There are more chance for human error -> Anyone can forget to set that statement to true again after changes.

We hope for answers or changes in a nearly future. 

Thank you so much in advance.

**Comment by aaronswainonset** [comment]
Is there any progress on this issue? We need to be able to control `prevent_destroy` for different target environments, and don't really want to duplicate code.

**Comment by JanisOrlovs** [comment]
This feature would be very useful. For example, Oracle cloud doesn't support deletion prevention. So we have to duplicate all the code.

**Comment by mattiabertorello** [comment]
+1 I've just run in this issue, please Hashicorp fix it.
I need to run terratest so with the `prevent_destroy = false` but in prod I want the property to `true`

**Comment by albertollamaso** [comment]
+1

**Comment by dan-geniusx** [comment]
+1, want to protect prod databases from deletion while allowing dev environments to freely delete their databases. Having to manually set `prevent_destroy = false` on dev environments creates the risk of setting `prevent_destroy = false` in prod.

**Comment by thiagolsfortunato** [comment]
+1, any news?

**Comment by AC90** [comment]
💥 🏃 😵 😬  just slammed into this brick wall as well. Any 🚑 coming soon Hashicorp ?

**Comment by luisamador** [comment]
Just came across this too :(

**Comment by bgshacklett** [comment]
I was able to work around this by creating per-environment [override files](https://www.terraform.io/language/files/override) which are copied into place as part of the deployment pipeline. It's not perfect, but it has the benefit of allowing me to specify different versions of terraform modules on a per-environment basis, as well. 

Example directory layout:
```
❯ tree
.
├── Jenkinsfile
├── README.md
├── iam.tf
├── locals.tf
├── main.tf
├── overrides
│   ├── dev
│   │   └── rds_override.tf
│   ├── qa1
│   ├── prod
│   └── uat
├── rds.tf
├── route53.tf
├── variables.tf
├── versions.tf
└── vpc.tf
```

I can't share the script that copies the files, but it uses the `find` command to find any override files under a certain path and the `exec` flag to copy them to the desired path.

**Comment by protosam** [comment]
Would also like to see support for `livecycle.prevent_destroy = var.A_STATIC_VAR`. My use-case was inside a module that uses the Github provider. My module supports figuring out if the user specified an existing repo or the repo needs to be created. I want to default this to "true", but permit users to override it with variables to the module for ephemeral environments.

**Comment by antonr-p2p** [comment]
WHY?!? 4 years to fix such a small issue!? WHY?? 
It was requested by so many people!

**Comment by antonr-p2p** [comment]
I face it still with Terraform v1.3.2 in 2022... really dissapointed

**Comment by Jasper-Ben** [comment]
> WHY?!? 4 years to fix such a small issue!? WHY?? It was requested by so many people!

Without having looked at the code, fixing such "small" issues might actually cascade into a massive amount of codebase rewrite, if hitting architectural limits. I'm pretty sure this is the case here, otherwise it would have been supported from the get-go 🙂

**Comment by rbjoergensen** [comment]
+1
Has Hashicorp given any reasoning as to why they're not fixing this?

**Comment by rubber-ant** [comment]
+1

**Comment by seed-doordash** [comment]
+1

**Comment by troupier88** [comment]
+1 We use terraform modules, the main dev set the default value at "true", that's not my use case :(

**Comment by crw** [comment]
Just a reminder to please use the 👍 reaction on the original post to upvote issues - we do sort by most upvoted to understand which issues are the most important. This also reduces "noise" in the notification feed for folks following this issue. Thanks!

@rbjoergensen 
> Has Hashicorp given any reasoning as to why they're not fixing this?

I believe the blocker is that to support this feature one would need to implement pre-processing of the configuration. This would be a major design change to the underlying fundamentals of Terraform. Not impossible, but not something that is likely to happen without a major product design effort. Again, please do not quote me on that technical explanation; this is how I understand the underlying issue but I may be a little off-base. 

You might also check out these adjacent issues:
https://github.com/hashicorp/terraform/issues/3116
https://github.com/hashicorp/terraform/issues/4149
https://github.com/hashicorp/terraform/issues/30937

**Comment by richardgragasin** [comment]
its 2023 guys, any updates?

**Comment by bgshacklett** [comment]
> I believe the blocker is that to support this feature one would need to implement pre-processing of the configuration.

Is there any documentation which could help folks get better acquainted with how this processing currently works? With a better understanding of the current difficulties/blockers, it would be easier to discuss potential solutions.

**Comment by karthik-twilio** [comment]
+1

**Comment by KamranBiglari** [comment]
+1

**Comment by crw** [comment]
Thanks for your interest in this issue! This is just a reminder to please avoid "+1" comments, and to use the upvote mechanism (click or add the 👍 emoji to the original post) to indicate your support for this issue. Thanks again for the feedback!

**Comment by algo7** [comment]
Any update on this one?

**Comment by sammalloyxydus** [comment]
+1 seems like a fairly common sense feature..

**Comment by denisp13** [comment]
I was surprised to find such a long and old tread for such a simple issue.... +1

**Comment by zeppelinen** [comment]
You are not alone

> I was surprised to find such a long and old tread for such a simple issue.... +1

**Comment by probststefan** [comment]
+1

**Comment by crw** [comment]
Thanks for your interest in this issue! This is just a reminder to please avoid "+1" comments, and to use the upvote mechanism (click or add the 👍 emoji to the original post) to indicate your support for this issue. Thanks again for the feedback!

**Comment by joe-a-t** [comment]
I want to call out that this is the root cause of a ton of other issues and work arounds that providers are either being asked to do or doing like:

- https://github.com/hashicorp/terraform/issues/24658
- https://github.com/hashicorp/terraform-provider-google/issues/10168
- https://github.com/hashicorp/terraform-provider-google/issues/7868
- https://github.com/hashicorp/terraform-provider-google/issues/2078
- https://github.com/hashicorp/terraform-provider-google/issues/13661

**Comment by scheleaap** [comment]
Mentioning #30957 here as it is related.

**Comment by thatsk** [comment]
any update on this .

**Comment by KevinPecquet** [comment]
Cannot use prevent_destroy in modules without a variable so for me it cannot be used without this feature.

**Comment by nsaccente** [comment]
Running Terraform v1.4.6, still no support for variables in `prevent_destroy` :upside_down_face:

**Comment by aaronswainonset** [comment]
Still can't use `prevent_destroy` effectively without support for variables. What's happening with it?

**Comment by ecoupal-believe** [comment]
+1

**Comment by michael-weiler** [comment]
any updates on this issue ?

**Comment by crw** [comment]
Thanks for the continued interest in this issue. As the use cases seem fairly clear and the replies primarily being of the +1 variety, I am going to lock comments to maintainers. For posterity, the last two substantive comments on this issue are: https://github.com/hashicorp/terraform/issues/22544#issuecomment-1301529658 and https://github.com/hashicorp/terraform/issues/22544#issuecomment-1507497841. Thanks again for your continued interest in this issue!
